proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,309 @@
+           34         if set(template_dict.keys()) != set(self.prompt_params):
+           35             available_params = set(list(template_dict.keys()) + list(set(kwargs.keys())))
+      ---> 36             raise ValueError(f"Expected prompt params {self.prompt_params} but got {list(available_params)}")
+           37
+           38         template_dict["prompt_template"] = self.prompt_text
+
+      ValueError: Expected prompt params ['documents'] but got ['some_unknown_param']
+  ```
+
+- MultiTaskPipelines
+
+  Even though we can use MultiTaskComponent directly its real power lies in using pipelines and Haystack. For example, we
+  can retrieve documents from the document store using the query and then inject the retrieved documents into documents
+  as a parameter to the selected MultiTaskComponent task. For example:
+
+  ``` python
+      from haystack.pipelines import MultiTaskPipeline
+      top_k = 3
+      query = "Who are the parents of Arya Stark?"
+      mtc = MultiTaskComponent(model_name_or_path="google/flan-t5-base")
+      pipe = MultiTaskPipeline(mtc, retriever)
+      output = pipe.run(query=query,
+                        params={"Retriever": {"top_k": top_k}},
+                        task="question-answering",
+                        questions=[query for n in range(0, top_k)],
+                        #documents parameter we need for this task will be automatically populated by the retriever
+                        )
+
+      output["results"]
+  ```
+
+  Or for example, question generation on retrieved documents:
+  ``` python
+      from haystack.pipelines import MultiTaskPipeline
+      top_k = 10
+      query = "Who are the parents of Arya Stark?"
+      pipe = MultiTaskPipeline(mtc, retriever)
+      output = pipe.run(query=query,
+                        task="question-generation",
+                        params={"Retriever": {"top_k": top_k}})
+      questions = output["results"]
+      questions
+  ```
+
+- MultiTaskPipeline chaining
+
+  Sometimes it is useful to save the results of the previous pipeline invocation and use that parameter in another
+  pipeline invocation for a specific NLP task. Let's use "question-answering" task with questions generated in
+  previous pipeline invocation for "question-generation".
+
+  ``` python
+      from haystack.pipelines import MultiTaskPipeline
+      top_k = 10
+      query = "Who are the parents of Arya Stark?"
+      pipe = MultiTaskPipeline(mtc, retriever)
+      output = pipe.run(query=query,
+                          task="question-generation",
+                          params={"Retriever": {"top_k": top_k}})
+      questions = output["results"]
+
+      output = pipe.run(query=query,
+                          task="question-answering",
+                          questions=questions,
+                          params={"Retriever": {"top_k": top_k}})
+      output["results"]
+  ```
```

Comment:
This part I personally don't like. 

In my mind this goes against the core concept of Pipeline: an object that lets you chain components one with the other. Why do I need several `MultiTaskPipeline` here, saving the intermediate inputs? I see this as a big flaw. `MultiTaskPipeline` should be either capable of handling several `MultiTaskComponent`, or it's not going to be very useful in practice.

Let me highlight the problem. What's better in your snippet than:

```python
top_k = 10
query = "Who are the parents of Arya Stark?"

docstore = InmemoryDocumentStore()
retriever = BM25Retriever(document_store=docstore)
mtc = MultiTaskComponent(model_name_or_path="google/flan-t5-base")

docs = retriever.retrieve(query=query, top_k=top_k)
questions = mtc( task="question-generation", query=query, documents=docs)
answers = mtc(task="question-answering", documents=docs, questions=questions)

print(answers)
```

I'd argue it's way neater without MultiTaskPipeline. It has one less retrieval step (saving time!) and it's easier to read. However, I value Pipelines a lot, so I believe these MultiTaskComponents would strongly benefit from being compatible with basic `Pipeline` objects.

Reply:
Resolved with https://github.com/deepset-ai/haystack/pull/3665/commits/835a8a690d9dcc4f31a30da03700bc4bcbd04e69